### ISSUE TITLE
ASNHC:Mamm - Bulkload Specimens

### ISSUE THREAD

**Issue by snknts** [ArctosDB/data-migration#147]
We (@diannakrejsa and myself) tried to bulk load 10 specimens on Thursday.  We thought they were loaded (or soon would be.. we set loaded to null)  but they've been sitting there since.

This morning, wanting to re-do the process before I ask for help here, I tried to "DELETE" them in the browsebulk.. and they're just sitting there with that status now..
Is there a step we're missing to commit/process those entries?

they're past the staging tables.. so clearing that doesn't work.. are we missing a role (and therefore a tool) maybe?

**Comment by Jegelewicz** [comment]
I used the SQL to delete them - I'm not sure why the "DELETE" in the AJAX grid wasn't doing anything, but it wasn't. They are gone now.

**Comment by snknts** [comment]
I'm pretty sure that's the same thing I did... any idea why it didn't work for me?

**Comment by snknts** [comment]
They seem to be stuck at DELETED again 
(they got duplicated, so I'm deleteing again)
am I not a data admin maybe?

**Comment by Jegelewicz** [comment]
Well, I guess it didn't work for me either! This is a question for @dustymc .BTW, just after deleting them when I returned to the AJAX grid, I was treated to the blue screen of death.

**Comment by snknts** [comment]
The deleted are gone now... maybe just a slow process?

I'll try to load them again :)

**Comment by snknts** [comment]
bulkload stager was happy..   bulkloader proper was not:
ORA-20001: Error and units, datum, and georeference source and protocol must be accompanied by coordinates

Our max error distance and units are empty.. is there a NULL-ish place holder we need to use instead?
(or do we need to load specimens with this data separately from those without? (instead of sending it empty?))

**Comment by diannakrejsa** [comment]
Yes good question! Do we need to sort by that?

**Comment by Jegelewicz** [comment]
You aren't required to provide an error, but if you do, you are expected to include units.  There is no placeholder. I don't understand how you could have something in max error, but have no units? Also, how can you have a max error if you have no coordinates in the first place? (Which is what this error is saying.)

**Comment by snknts** [comment]
We don't in this file.. but we do for other records so things are written to include it.
i.e. if we load 100 records and 1 has error distance and units...
In the file, the 99 would have null values in those columns.. the 1 would have real values. 
in this case all ten don't.. so there all null.
(does that make sense?)

**Comment by Jegelewicz** [comment]
Not really... Null should not pose a problem, are those without truly NULL or do they end up containing " "? I don't see anything sitting in the bulkloader, so I can't tell what is really going on.

**Comment by snknts** [comment]
in the csv they're empty..
i.e. ,,
I'll load them again momentariy :)

**Comment by snknts** [comment]
They're at the same state they were yesterday.  I attached 
[mammalogy-Base-20191031-ASNHC1-10.zip](https://github.com/ArctosDB/data-migration/files/3804962/mammalogy-Base-20191031-ASNHC1-10.zip)
the csv file too (zipped)

**Comment by Jegelewicz** [comment]
The issue is that you cannot have data in any of these:

ORIG_LAT_LONG_UNITS | DEC_LAT | DEC_LONG | DATUM | GEOREFERENCE_SOURCE | MAX_ERROR_DISTANCE | MAX_ERROR_UNITS | LOCALITY_REMARKS | GEOREFERENCE_PROTOCOL | MAXIMUM_ELEVATION | MINIMUM_ELEVATION | ORIG_ELEV_UNITS

if you don't have coordinates.

**Comment by Jegelewicz** [comment]
But you have:


ORIG_LAT_LONG_UNITS | DEC_LAT | DEC_LONG | DATUM | GEOREFERENCE_SOURCE | MAX_ERROR_DISTANCE | MAX_ERROR_UNITS | LOCALITY_REMARKS | GEOREFERENCE_PROTOCOL | MAXIMUM_ELEVATION | MINIMUM_ELEVATION | ORIG_ELEV_UNITS
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  
decimal   degrees |   | World Geodetic System 1984 | unknown |   |   |   | not recorded |   |  

All of those "decimal degrees, WGS 1984, unknown, and not recorded need to go!

**Comment by snknts** [comment]
ok.. I bet we're including some placeholders for records that we should send as blank.  I'll update our stuff and try again.

Thank you for the help :)

**Comment by campmlc** [comment]
Make sure you don't have anything in the columns, including spaces. Oracle
will consider spaces as content.

**Comment by snknts** [comment]
Success! 
:)

**Comment by snknts** [comment]
follow up question ;)

a few of or specimens don't have any dates assigned.. and it seems arctos requires an EVENT_ASSIGNED_DATE and ATTRIBUTE_DATE... 
null isn't allowed in its place.. what should we do in this case?

**Comment by Jegelewicz** [comment]
This is up to you, but I always encourage collections to apply some consistent rule and document it for yourself so that those who come after you can figure it out.

For Attribute dates, I use the preparation date if known or the collecting date if known. If neither are known, see below.

For specimen event dates, use the date of georeference if known or the collection date if known. If you have absolutely no dates associated with a record, the simplest method is to use a generic date that won't conflict with any of your known dates "1800-01-01" is a favorite. More time consuming, but more rewarding later is to make a best guess based upon the age of the collector, known information about your collection (for example -  nothing was collected before 1970) or some other method.

Let me know if that doesn't help!

**Comment by snknts** [comment]
made it past those.. and have a two new issues on the next batch :)
one was another all or nothing issue.. I believe I have that resolved...
the other is more mysterious:
ORA-00001: unique constraint (UAM.IX_COLLECTOR_COID_AID_ROLE) violated
I'm not sure what the issue is there...

**Comment by Jegelewicz** [comment]
I've never seen that one! It will be something in the collector roles I'm guessing. But I see what it is - you have the same agent listed as collector twice. Remove the agent and role from COLLECTOR2 and it should work.

**Comment by snknts** [comment]
ahh makes sense.  thanks again :)

**Comment by snknts** [comment]
We found that a lot of our collection objects have bad lat/long values..
usually just sign issues (i.e. a longitude being +100 instead of -100 (east and west being swapped))

I can't seem to correct our records though.
(I get a "You are not authorized to access this form.") error when I try to get to the locaility.

Is this something I can change?

**Comment by Jegelewicz** [comment]
Sure, but I want to make sure that you understand the shared nature of localities.

Please read and acknowledge that you understand the following:

[Understanding Arctos Localities](http://handbook.arctosdb.org/documentation/locality.html)

[Understanding the Arctos Locality Model](http://handbook.arctosdb.org/how_to/How-to-understand-the-Arctos-Locality-Model.html)

**Comment by diannakrejsa** [comment]
Hi Teresa, I've read through the materials you linked above. Am I correct in understanding that if one created lat/longs for a record that shared locality with other records within ASNHC or all of Arctos, it would change all of them? I'd like to start just by georeferencing individual records or localities that contain one record. We have a number of those.

**Comment by Jegelewicz** [comment]
> Am I correct in understanding that if one created lat/longs for a record that shared locality with other records within ASNHC or all of Arctos, it would change all of them?

Yep! Which can be either awesome or a drag. Starting with a locality that involves only one specimen record that is in your collection is an excellent idea - it will let you see how the the model works on a very small scale.

Are you planning to add coordinates to the collecting event or directly to the locality? If you want, we can set up a Zoom and walk through one together.

**Comment by diannakrejsa** [comment]
Yes, I can see how it could be great and time saving or really bad, haha.

I believe I would go the route of changing direct to locality, then refresh the list for ASNHC records left without coordinate data. But, this may not be the best way to operate. A Zoom session would be great! I'm free all day, haha.

**Comment by Jegelewicz** [comment]
@diannakrejsa sorry to just get back here! How about we meet up tomorrow at 9AM MDT? We can use my Zoom https://us04web.zoom.us/j/7482621548

**Comment by diannakrejsa** [comment]
Sounds good! I remarkably have another Zoom at 10AM MDT, so I won't take too much of your time. Thanks, talk tomorrow!
